hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java

```diff
@@ -753,10 +755,47 @@ protected void checkPath(Path path) {
           return;
       }
     }
+
+    // The authority of the input path does not match with the
+    // authority of the client. Check if it matches with any
+    // of the available NNs in the system.
+    String nns = getConf().get("dfs.namenodes.rpc.addresses");
+    if (nns != null) {
+      uri = canonicalizeUri(uri);
+      String pathAuthority = uri.getAuthority();
+      StringTokenizer st = new StringTokenizer(nns, ",");
+      while (st.hasMoreTokens()) {
+        String nnAuthority = getNNAuthority(st.nextToken(), thisScheme);
+        if (nnAuthority != null && nnAuthority.equalsIgnoreCase(pathAuthority)) {
+          return;
+        }
+      }
+    }
+
     throw new IllegalArgumentException("Wrong FS: " + path +
                                        ", expected: " + this.getUri());
   }
 
+  private String getNNAuthority(String nn, String thisScheme) {
+    try {
+      if(!nn.contains("://")){ 
+        return nn;
+      } else {
+        // if the scheme is set then make sure that the scheme matches
+        URI uri = URI.create(nn);
+        String nnScheme = uri.getScheme();
+        if (thisScheme.equalsIgnoreCase(nnScheme)) {
+          URI nnURI = canonicalizeUri(uri);
+          String nnAuthority = nnURI.getAuthority();
+          return nnAuthority;
+        }
+      }
+    } catch (NullPointerException | IllegalArgumentException e) { // URI creation failed
+      LOG.warn("Make sure that dfs.namenodes.rpc.addresses is properly configured in core-site.xml");
+    }
+    return null;
+  }
+
```

Comment:
I think this should be in the subclass DistributedFileSystem and not in here because there are other implementation of FileSystem that are not compatible with this code.

Reply:
I have moved it to the DFS class